### PR TITLE
Add Competence components to frontend

### PR DESCRIPTION
**Pull request by NealGeilen**
Adding the basic components to the frontend competence document

### CODE REVIEW

File: Epsilon.Host.WebApi/Controllers/LearningController.cs [hunk 1]
```diff
@@ -37,4 +37,18 @@ public async Task<ActionResult<LearningDomain>> GetDomain(string name)
 
         return Ok(domain);
     }
+    
+    
+    //TODO filter on tenetID and DomainID
+    [HttpGet("domain/{tenetId}/outcomes")]
```
Comment (Typiqally) [inline Epsilon.Host.WebApi/Controllers/LearningController.cs:43]: tenet should be tenant, also we should prevent commiting TODO stuff and instead move it to a separate PR. to keep the repo clean.

File: Epsilon.Abstractions/LearningDomainCriteria.cs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Epsilon.Abstractions;
+
+public record LearningDomainCriteria(
+    int Id,
+    double? MasteryPoints
```
Comment (Typiqally) [inline Epsilon.Abstractions/LearningDomainCriteria.cs:5]: Should probably be in outcome result

File: Epsilon.Abstractions/LearningDomainOutcomeResult.cs [hunk 3]
```diff
@@ -2,8 +2,5 @@ namespace Epsilon.Abstractions;
 
 public record LearningDomainOutcomeResult(
     LearningDomainOutcome Outcome,
-    double Grade,
-    DateTime SubmittedAt,
-    string Assignment,
-    Uri AssignmentUrl
```
Comment (Typiqally) [inline Epsilon.Abstractions/LearningDomainOutcomeResult.cs:8]: Can you give reason for removing these?

File: Epsilon.Abstractions/LearningDomainOutcomeResult.cs [hunk 4]
```diff
@@ -2,8 +2,5 @@ namespace Epsilon.Abstractions;
...
+    double? Grade
```
Comment (Typiqally) [inline Epsilon.Abstractions/LearningDomainOutcomeResult.cs:5]: Can grade be null? Shouldn't be null imo, but this is probably because of the kpi matrix?

File: Epsilon.Host.Frontend/src/components/Competence/CompetenceProfile.vue [hunk 5]
```diff
@@ -0,0 +1,77 @@
+<script setup lang="ts">
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/components/Competence/CompetenceProfile.vue:1]: As per Vue's conventions, template should come first, then script, then styling

File: Epsilon.Host.Frontend/src/components/Competence/KpiMatrix.vue [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import KpiMatrixCell from "@/components/Competence/KpiMatrixCell.vue"
+
+const store = useStore()
+
+const allOutcomes = computed(
+    () =>
+        store.state.filterdSubmissions
+            .flatMap((sub) => sub.criteria?.map((cr) => cr.id as string))
+            .filter((value, index, self) => self.indexOf(value) === index)
+    // .map((uniqueOutcome) => props.domain.)
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/components/Competence/KpiMatrix.vue:48]: Remove commented code

File: Epsilon.Host.Frontend/src/components/Competence/KpiMatrixCell.vue [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+}>()
+
+function getColor(): string {
+    if (
+        props?.result?.grade != null &&
+        props?.criteria?.masteryPoints != null
+    ) {
+        return (((props?.result?.grade as unknown as number) >=
+            props?.criteria?.masteryPoints) as unknown as number)
+            ? "#44F656"
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/components/Competence/KpiMatrixCell.vue:16]: Hardcoded color values

File: Epsilon.Host.Frontend/src/components/TopNavigation.vue [hunk 8]
```diff
@@ -1,80 +1,78 @@
...
+const store = useStore()
 const selectedTerm: Ref<EnrollmentTerm | undefined> = ref(undefined)
 
-onMounted(() => {
-    api?.filter.accessibleStudentsList().then((r: HttpResponse<User[]>) => {
-        students.value = r.data
-        selectedStudent.value = students.value[0]
-    })
+watch(selectedTerm, () => {
+    console.log("Term selection")
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/components/TopNavigation.vue:39]: Remove this log

File: Epsilon.Host.Frontend/src/components/TopNavigation.vue [hunk 9]
```diff
@@ -90,18 +88,18 @@ const avatarImage = computed(() => {
...
+//        max-width: 3rem;
+//        max-height: 3rem;
+//        aspect-ratio: 1/1;
+//        border: none;
+//        border-radius: 3rem;
+//        margin-right: 1rem;
+//        overflow: hidden;
+//        background-color: #fff;
+//    }
+//}
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/components/TopNavigation.vue:104]: Commented code, also why remove this?

File: Epsilon.Host.Frontend/src/views/CompetenceDocument.vue [hunk 10]
```diff
@@ -1,3 +1,11 @@
 <template>
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/views/CompetenceDocument.vue:1]: Just beware that by using this store approach it becomes very difficult to see where data is coming from. I don't think the components itself should incorporate the store variables, instead they can be used in the properties, making it clear where the data is coming from and allowing for more scalability in case we switch from store type.

File: Epsilon.Host.Frontend/src/views/DashboardView.vue [hunk 11]
```diff
@@ -24,14 +24,40 @@
         </main>
     </TabGroup>
 </template>
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/views/DashboardView.vue:26]: Nothing big, but please add a seperatory enter here

File: Epsilon/Services/LearningOutcomeCanvasResultService.cs [hunk 12]
```diff
@@ -53,23 +61,18 @@ ILearningDomainService learningDomainService
...
-    public async IAsyncEnumerable<LearningDomainOutcomeResult> GetOutcomeResults(string studentId)
+    public async IAsyncEnumerable<LearningDomainSubmission> GetSubmissions(string studentId)
     {
         var submissionsTask = _canvasGraphQlApi.Query(Query, new Dictionary<string, object> { { "studentIds", studentId }, });
         var domainOutcomesTask = _learningDomainService.GetOutcomes();
 
         await Task.WhenAll(submissionsTask, domainOutcomesTask);
 
-        foreach (var submission in submissionsTask.Result.Courses.SelectMany(static c => c.Submissions.Nodes))
+        foreach (var submission in submissionsTask?.Result?.Courses?.SelectMany(static c => c.Submissions.Nodes))
```
Comment (Typiqally) [inline Epsilon/Services/LearningOutcomeCanvasResultService.cs:71]: We should really do some null checks instead of adding these lazy ?'s Otherwise it will become a pain to debug in later versions, trust me on this one.

**Review by Typiqally — CHANGES_REQUESTED** [review]
Large commit means a lot of comments, sorry bout that. No major changes required, just some questions and things to keep in mind. But please make sure to keep everything clean before merging as to prevent any technical debt build up from before. It will bite u in the ass and I can tell from experience that it's not nice.

File: Epsilon.Abstractions/LearningDomainOutcomeResult.cs (hunk 4 above)
Comment (NealGeilen) [reply Epsilon.Abstractions/LearningDomainOutcomeResult.cs:5]: Correct, When not graded, value will be null. And is needed for KPI matrix

File: Epsilon.Host.Frontend/src/views/CompetenceDocument.vue (hunk 10 above)
Comment (NealGeilen) [reply Epsilon.Host.Frontend/src/views/CompetenceDocument.vue:1]: Well oke... But i need to export to mention it in `main.ts`

File: Epsilon.Host.Frontend/src/components/Competence/KpiMatrixCell.vue (hunk 7 above)
Comment (NealGeilen) [reply Epsilon.Host.Frontend/src/components/Competence/KpiMatrixCell.vue:16]: Correct,  They indicate the stage of the KPI (Graded/Submitted/Masterd) in a given assignment for the KPI matrix. 
In my view its useless to make that dynamic.

File: Epsilon.Host.Frontend/src/components/Competence/CompetenceProfile.vue (hunk 5 above)
Comment (NealGeilen) [reply Epsilon.Host.Frontend/src/components/Competence/CompetenceProfile.vue:1]: Oke, but the there needs to be a lint check. Beacuse Webstrom creates it in this order

Comment (NealGeilen) [reply Epsilon.Host.Frontend/src/components/Competence/CompetenceProfile.vue:1]: Rules added 
https://eslint.vuejs.org/rules/component-tags-order.html

File: Epsilon.Host.Frontend/src/components/TopNavigation.vue (hunk 9 above)
Comment (NealGeilen) [reply Epsilon.Host.Frontend/src/components/TopNavigation.vue:104]: Because it was determined that getting pictures from canvas was a huge bottleneck in loading times.

File: Epsilon.Abstractions/LearningDomainCriteria.cs (hunk 2 above)
Comment (NealGeilen) [reply Epsilon.Abstractions/LearningDomainCriteria.cs:5]: When items are not graded, there is not a result generated. This is thus needed for the KPI Matrix

File: Epsilon.Abstractions/LearningDomainOutcomeResult.cs (hunk 3 above)
Comment (NealGeilen) [reply Epsilon.Abstractions/LearningDomainOutcomeResult.cs:8]: Because it's moved to `LearningDomainSubmission`  The parent object.

File: Epsilon/Services/LearningOutcomeCanvasResultService.cs (hunk 12 above)
Comment (NealGeilen) [reply Epsilon/Services/LearningOutcomeCanvasResultService.cs:71]: True, but in the end it's a disadvantage of the architecture and  GraphQL. 
With in the same file it's determined what can be null and not. 

Those checks are only needed within the services that parse GraphQL data to our own models.

File: Epsilon.Abstractions/LearningDomainCriteria.cs (hunk 2 above)
Comment (Typiqally) [reply Epsilon.Abstractions/LearningDomainCriteria.cs:5]: Valid reason

File: Epsilon.Abstractions/LearningDomainOutcomeResult.cs (hunk 3 above)
Comment (Typiqally) [reply Epsilon.Abstractions/LearningDomainOutcomeResult.cs:8]: Valid

File: Epsilon.Abstractions/LearningDomainOutcomeResult.cs (hunk 4 above)
Comment (Typiqally) [reply Epsilon.Abstractions/LearningDomainOutcomeResult.cs:5]: Should probably not be an outcome result then maybe? As this is usually where grades are set. Because this is for individual outcomes

File: Epsilon.Host.Frontend/src/components/Competence/CompetenceProfile.vue (hunk 5 above)
Comment (Typiqally) [reply Epsilon.Host.Frontend/src/components/Competence/CompetenceProfile.vue:28]: 👍🏼

File: Epsilon.Host.Frontend/src/components/Competence/KpiMatrixCell.vue (hunk 7 above)
Comment (Typiqally) [reply Epsilon.Host.Frontend/src/components/Competence/KpiMatrixCell.vue:16]: We should probably combine this with the regular colors maybe? Dunno but we probably should not hardcode it because same values are used in word generation

File: Epsilon.Host.Frontend/src/components/TopNavigation.vue (hunk 9 above)
Comment (Typiqally) [reply Epsilon.Host.Frontend/src/components/TopNavigation.vue]: No need to remove the avatar styling, just use a default character for now and put the feature on the roadmap

File: Epsilon/Services/LearningOutcomeCanvasResultService.cs (hunk 12 above)
Comment (Typiqally) [reply Epsilon/Services/LearningOutcomeCanvasResultService.cs:71]: For this instance I agree, but watch out for other values that CAN be null even in GraphQL response such as grade.

File: Epsilon.Host.Frontend/src/components/Competence/KpiMatrixCell.vue (hunk 7 above)
Comment (NealGeilen) [reply Epsilon.Host.Frontend/src/components/Competence/KpiMatrixCell.vue:16]: Oke, But if we are going to follow that logic. Then there needs to be a new endpoint for these color codes. Because they don't fit within the domain models. And are not part of the results.

File: Epsilon.Abstractions/LearningDomainOutcomeResult.cs (hunk 4 above)
Comment (NealGeilen) [reply Epsilon.Abstractions/LearningDomainOutcomeResult.cs:5]: True when following that logic the name is incorrect

File: Epsilon.Host.Frontend/src/components/TopNavigation.vue (hunk 9 above)
Comment (NealGeilen) [reply Epsilon.Host.Frontend/src/components/TopNavigation.vue]: I get it. But is is in the git history. 
And especially for CSS the less is in the project the better. 
the less stuff it can mess up.

**Review by Typiqally — APPROVED** [review]
